### PR TITLE
devenv: assert --impure is used

### PR DESCRIPTION
**Pull request by bobvanderlinden**
Currently it is possible to run devenv without `--impure`. However, strange things may happen, because `$PWD` is not set. This results in `env.DEVENV_ROOT` being `""`, which results in various other variables to contain an invalid path (like `/.devenv`).
This will go unnoticed, but can cause problems.

This PR makes fetchign `env.DEVENV_ROOT` throw an error whenever it is fetched without Nix running in `--impure` mode.

I can imagine this can also cause annoyances for people who are using devenv without _actually_ depending on `DEVENV_ROOT`. So it isn't a clear win in that regard. I thought I'd just suggest the change.

This was in response to https://github.com/cachix/devenv/issues/550

### CODE REVIEW

**Comment by domenkozar** [comment]
Some weird CI errors popped up.

**Comment by bobvanderlinden** [comment]
I see. Strange indeed. It seems the `system` attribute isn't passed by `callPackage`, `pkgs.system` doesn't exist?
It _seems_ it is only happening on the self-hosted MacOS. Does it have a different version of Nix?

That said, I can work around it by manually supplying `system` using `callPackage { system = pkgs.stdenv.system; }` or something alike. Would that be a good solution?

**Comment by domenkozar** [comment]
That's a good solution, since `pkgs.system` is deprecated and we're using `pkgs.stdenv.system` where appropriate.

File: src/modules/top-level.nix [hunk 1]
```diff
@@ -104,7 +104,18 @@ in
 
   config = {
     # TODO: figure out how to get relative path without impure mode
-    env.DEVENV_ROOT = builtins.getEnv "PWD";
+    env.DEVENV_ROOT =
+      let
+        pwd = builtins.getEnv "PWD";
+      in
+      if pwd == null then
```
Comment (domenkozar) [inline src/modules/top-level.nix:111]:
```suggestion
      if pwd == "" then
```

Comment (bobvanderlinden) [reply src/modules/top-level.nix]: Good call. Fixed this. Also found a way to test this properly, which also catches this :+1:

**Comment by bobvanderlinden** [comment]
I made mkNakedShell use `pkgs.stdenv.system` instead of `pkgs.system`, added a test that checks whether nix-develop fails in pure mode, lastly fixed the getEnv resulting in `""` instead of `null`.

**Pull request merged**